ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -109,11 +109,27 @@ module LocalFlow {
    * Holds if `nodeFrom` is a parameter node, and `nodeTo` is a corresponding SSA node.
    */
   predicate localFlowSsaParamInput(Node nodeFrom, Node nodeTo) {
-    nodeTo = getParameterDefNode(nodeFrom.(ParameterNode).getParameter())
+    nodeTo = getParameterDefNode(nodeFrom.(ParameterNodeImpl).getParameter())
     or
     nodeTo = getSelfParameterDefNode(nodeFrom.(SelfParameterNode).getMethod())
   }
 
+  /**
+   * Holds if `nodeFrom -> nodeTo` is a step from a parameter to a capture entry node for
+   * that parameter.
+   *
+   * This is intended to recover from flow not currently recognised by ordinary capture flow.
+   */
+  predicate localFlowSsaParamCaptureInput(Node nodeFrom, Node nodeTo) {
```

Comment:
The underlying bug is that SSA doesn't consider a lambda expression to escape when returned, as highlighted by the added test case. As mentioned in the PR description, the code here is a workaround to avoid regression in the short-term, and I'm not really interested in generalizing the workaround. I initially tried landing such fixes prior to opening this PR, but after a while I didn't want to block the PR any longer.

Obviously the bug in SSA should be fixed, but at the architectural level I don't think this layer ought care about it. I'd like to post-pone flow-sensitive handling of captured variables until after the call graph, so the layers would look something like:
- Materialize data-flow nodes.
- Compute flow-insensitive step relation, treating captured variables flow-insensitively (basically SSA only used for true locals)
- Compute call graph, using the flow-insensitive relation for propagating type information
- Compute flow-sensitive step relation (for use by DataFlow::Configuration etc). This stage can then in principle benefit from seeing the call graph.

Flow sensitivity doesn't matter very much for propagating type information, whereas it matters a lot more to taint tracking and the like. Using a single handling of captured variables to cover both isn't optimal.

I tried to make a start on this [here](https://github.com/github/codeql/pull/11010) but as mentioned, I had to shelf it to make some progress on this PR.

Reply:
Makes perfect sense; thanks.